### DISCUSSION TITLE
Which of these algorithms would be a good contribution to networkx?

### DISCUSSION THREAD

**Discussion by erelsgl** [networkx/networkx#5164] [Algorithms]
Some of my talented students would like to contribute some algorithms to `networkx` as their undergraduate project. Which of the following algorithms would you consider a good contribution?

1. **Priority maximum matching**: a matching that, among all maximum matchings, gives priorities to some vertices over others. Useful in applications such as kidney exchange. There are polynomial-time algorithms for [general graphs](https://arxiv.org/abs/1512.08555) and [bipartite graphs](https://arxiv.org/abs/1512.09349).

2. **Cycle packing**: a largest collection of disjoint cycles of size at most k. Useful for kidney exchange. The case k=2 corresponds to matching. For k>2 the problem is NP-hard but has some approximation algorithms: [paper 1](https://www.worldscientific.com/doi/abs/10.1142/S1793830909000373), [Paper 2](https://www.pnas.org/content/112/3/663.short).

3. **Maximum envy-free matching**: a matching in a bipartite graph (X+Y, E), in which an unmatched vertex in X is not adjacent to any matched vertex in Y. It is a useful subroutine for various fair division algorithms. [Polynomial-time algorithm](https://arxiv.org/abs/1901.09527)

4. **Fair division of a graph**: dividing a graph to connected components satisfying some fairness conditions. [Polynomial-time algorithm in section 5](https://arxiv.org/abs/1705.10239).

5. **Rank-maximal matching**: a matching in a bipartite graph with sides X,Y. It maximizes the number of vertices in X that are linked to their first priority; subject to that, the number of vertices linked to their second priority; etc. [Polynomial-time algorithm](https://dl.acm.org/doi/10.1145/1198513.1198520).

6. **Smallest multiway cut**: Given some k vertices, find the smallest (or lowest-weight) cut that separates among these vertices. The problem of minimum-cut is the special case k=2. For k>2 it is NP-hard, but there are [approximation algorithms](https://dl.acm.org/doi/pdf/10.1145/276698.276711).

7. **Fair flow**: computing a network flow that satisfies some fairness properties. [Paper](https://arxiv.org/abs/1907.02673).

8. **Rainbow matching**: each edge has a color; the goal is to find a largest matching in which each edge has a different color. The problem is NP hard but has some [approximation algorithms](https://link.springer.com/article/10.1007/s00453-018-0497-3).

9. **Dominating set**: networkx already has some algorithms, but maybe some other algorithms can be useful too? [Wikipedia page](https://en.wikipedia.org/wiki/Dominating_set#Algorithms_and_computational_complexity).

10. **Contiguous oriented labeling**:  a labeling of the vertices of a graph with consecutive integers, such that every prefix or suffix is connected. [Polynomial-time algorithm for almost-bridgeless graphs](https://arxiv.org/pdf/1910.14129.pdf#subsection.4.1).

11. **[Fractional matching](https://en.wikipedia.org/wiki/Fractional_matching)**: a matching in which each edge can participate fractionally. There are algorithms for finding a maximum-weight fractional matching, as well as decomposing such a matching into integral matchings using [Birkhoff's algorithm](https://en.wikipedia.org/wiki/Birkhoff_algorithm).

12. **Fair dissections**: partitioning a graph into some k connected components, such that the sum of weights in each component is as close as possible to the average weight. In general it is NP-hard, but there are [polynomial-time algorithms for special cases](https://onlinelibrary.wiley.com/doi/abs/10.1002/net.3230200305).

**Comment by dschult** [comment]
These each look good — though fairly specialized. Also, be careful that a paper may give a polynomial time algorithm that is incredibly slow if the paper is solely interested in the asymptotic time behavior. It’s probably good for the students to have a crude naive algorithm to compare their final results to.  Even just having to check the timing helps the students be aware that asymptotic results can still have a large constant multiplier and thus not be practical. 

These look like fun — I don’t think I have an easy way to pick between them.

**Reply by LielVaknin** [reply]
Hey Dan,
We are Erel's students, we decided to continue with the fifth algorithm ["Rank maximal matchings"](https://dl.acm.org/doi/10.1145/1198513.1198520).
We wanted to consult with you about the output of the algorithm. 
Should it be a matching dictionary like similar algorithms or should it be a matching dictionary in addition to a list where the ith element is the number of people who get their ith preference, for example: [2,1] - 2 people get their first choice and only 1 person get his second choice. 
At first we wrote the signatures of the algorithm methods as you can see [here](https://github.com/LielVaknin/NetworkX-final-project/blob/master/Rank-Maximal-Matchings%20algorithm/Initial%20Code%20And%20Tests/rank_maximal_matching.py).
In addition, you can see an example of the algorithm [here](https://github.com/LielVaknin/NetworkX-final-project/blob/master/Rank-Maximal-Matchings%20algorithm/Summary%20and%20examples/Rank%20Maximal%20Matchings%20-%20example.pdf).
We would like to hear your opinion,
Liel, Amiel and Oriya.

**Reply by dschult** [reply]
It seems that from the matching (and already having the rank stored in the graph) that the user can get all the desired information. For example, the number of people who get each rank can come using `Counter`:

```python
counts_of_ranks = collections.Counter(G[node][matching[node]]["weight"] for node in G)
for i in counts_of_ranks:
    print(f"{counts_of_ranks[i]} people got choice number {i}.")
```

Also, a dict keyed by node to the rank matched can be constructed using dict comprehension:
```python
rank_by_node = {node: G[node][matching[node]]["weight"] for node in G}
```

The signature looks reasonable, but I would add an argument `weight` or maybe `rank` to indicate the attribute name for the attribute on the edge that holds the rank.  I'm guessing your examples use "weight", but this would make it easy for someone to store it under a different name. 
:}

**Reply by LielVaknin** [reply]
Hi Dan,
We made a pull request for the first algorithm "Rank Maximal Matching" as you can see [here](https://github.com/networkx/networkx/pull/5585).
In addition, we want to start working on another algorithm.
We thought about [Birkhoff algorithm](https://en.wikipedia.org/wiki/Birkhoff_algorithm).
We would like to hear your opinion,
Liel, Amiel and Oriya.

**Comment by dschult** [comment]
Thanks for the PR! :}

I am not familiar with the Birkhoff algorithm and the Wikipedia link shows that there are both old and recent papers exploring it. So that sounds like a good algorithm to start working on.
Enjoy!

**Comment by erelsgl** [comment]
I see that currently there is no algorithm for finding a maximum fractional matching. Would this be a good addition to `networkx`?

**Comment by erelsgl** [comment]
Hi Dan. Here is a pull request for number 3) in the list: https://github.com/networkx/networkx/pull/6344
Can you please look at it?

**Comment by PurviChaurasia** [comment]
Hi! This seems like an extremely interesting discussion. I was wondering if I could help out with building these algorithms in any way @erelsgl ? It's amazing to see how you're guiding your students in this field. Please let me know if I can join in and work with all of you as well :)

**Reply by erelsgl** [reply]
Sure! You are welcome to work on algorithms.

**Comment by hoehrmann** [comment]
If I may add some wishlist items...

»**Linear-time modular decomposition of directed graphs**« - https://doi.org/10.1016/j.dam.2004.02.017
»**An improved algorithm for hierarchical clustering using strong components**« - https://doi.org/10.1016/0020-0190(83)90088-1
